kernel/src/sched/sched_class/mod.rs

```diff
@@ -0,0 +1,320 @@
+        let cur = guard.current_cpu();
+        if affinity.contains(cur) {
+            cur
+        } else {
+            affinity.iter().next().expect("empty affinity")
+        }
+    }
+}
+
+impl PerCpuClassRqSet {
+    fn pick_next_thread(&mut self) -> Option<Arc<Thread>> {
+        (self.stop.pick_next())
+            .or_else(|| self.real_time.pick_next())
+            .or_else(|| self.fair.pick_next())
+            .or_else(|| self.idle.pick_next())
+    }
+
+    fn enqueue_thread(&mut self, thread: &Arc<Thread>) {
+        let attr = thread.sched_attr();
+
+        attr.period_start.store(sched_clock(), Relaxed);
+        attr.start.store(sched_clock(), Relaxed);
+
+        let cloned = thread.clone();
+        match *attr.policy.lock() {
+            SchedPolicy::Stop => self.stop.enqueue(cloned),
+            SchedPolicy::RealTime { .. } => self.real_time.enqueue(cloned),
+            SchedPolicy::Fair(_) => self.fair.enqueue(cloned),
+            SchedPolicy::Idle => self.idle.enqueue(cloned),
+        };
+    }
+}
+
+impl LocalRunQueue for PerCpuClassRqSet {
+    fn current(&self) -> Option<&Arc<Task>> {
+        self.current.as_ref()
+    }
+
+    fn pick_next_current(&mut self) -> Option<&Arc<Task>> {
+        match self.pick_next_thread() {
+            Some(next) => {
+                let next_task = next.task();
+                if let Some(old_task) = self.current.replace(next_task.clone()) {
+                    if Arc::ptr_eq(&old_task, &next_task) {
+                        return None;
+                    }
+                    let old = Thread::borrow_from_task(&old_task);
+                    self.enqueue_thread(old);
+                }
+                self.current.as_ref()
+            }
+            None => None,
+        }
+    }
+
+    fn update_current(&mut self, flags: UpdateFlags) -> bool {
+        if let Some(cur_task) = &self.current {
+            let cur = Thread::borrow_from_task(cur_task);
+            let now = sched_clock();
+
+            let (update, lookahead) = match &*cur.sched_attr().policy.lock() {
+                SchedPolicy::Stop => (self.stop.update_current(now, cur, flags), 0),
+                SchedPolicy::RealTime { .. } => (self.real_time.update_current(now, cur, flags), 1),
+                SchedPolicy::Fair(_) => (self.fair.update_current(now, cur, flags), 2),
+                SchedPolicy::Idle => (self.idle.update_current(now, cur, flags), 3),
+            };
+
+            cur.sched_attr().start.store(now, Relaxed);
+
+            let lookahead = {
+                let mut ret = false;
+                if lookahead >= 1 {
+                    ret |= !self.stop.is_empty();
+                }
+                if lookahead >= 2 {
+                    ret |= !self.real_time.is_empty();
+                }
+                if lookahead >= 3 {
+                    ret |= !self.fair.is_empty();
+                }
+                ret
+            };
+            update || lookahead
```

Comment:
First of all, `update` is a poorly-chosen word. You can do a better job on naming.

Second, when `update` is true, we can simply short circuit the logic of "lookahead".